javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+ */
+package javaslang.collection;
+
+import javaslang.Function1;
+import javaslang.Tuple;
+import javaslang.Tuple2;
+import javaslang.Tuple3;
+import javaslang.control.Option;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.*;
+import java.util.stream.Collector;
+
+import static javaslang.collection.Comparators.naturalComparator;
+
+/**
+ * An immutable {@code BitSet} implementation.
+ *
+ * @author Ruslan Sennov
+ * @since 2.1.0
+ */
+public interface BitSet<T> extends SortedSet<T> {
+
+    long serialVersionUID = 1L;
+
+    class Builder<T> {
+
+        final static Builder<Integer> DEFAULT = new Builder<>(i -> i, i -> i);
+
+        final Function1<Integer, T> fromInt;
+        final Function1<T, Integer> toInt;
+
+        Builder(Function1<Integer, T> fromInt, Function1<T, Integer> toInt) {
+            this.fromInt = fromInt;
+            this.toInt = toInt;
+        }
+
+        public Collector<T, ArrayList<T>, BitSet<T>> collector() {
+            final BinaryOperator<ArrayList<T>> combiner = (left, right) -> {
+                left.addAll(right);
+                return left;
+            };
+            return Collector.of(ArrayList::new, ArrayList::add, combiner, this::ofAll);
+        }
+
+        public BitSet<T> empty() {
+            return new BitSetModule.BitSet1<>(fromInt, toInt, 0L);
+        }
+
+        public BitSet<T> of(T t) {
+            final int value = toInt.apply(t);
+            if(value < BitSetModule.BITS_PER_WORD) {
+                return new BitSetModule.BitSet1<>(fromInt, toInt, 1L << value);
+            } else if(value < 2 * BitSetModule.BITS_PER_WORD) {
+                return new BitSetModule.BitSet2<>(fromInt, toInt, 0L, 1L << value);
+            } else {
+                return empty().add(t);
+            }
+        }
+
+        @SuppressWarnings("varargs")
+        @SafeVarargs
+        public final BitSet<T> of(T... values) {
+            return empty().addAll(Array.wrap(values));
+        }
+
+        public BitSet<T> ofAll(Iterable<? extends T> values) {
+            Objects.requireNonNull(values, "values is null");
+            return empty().addAll(values);
+        }
+
+        public BitSet<T> tabulate(int n, Function<? super Integer, ? extends T> f) {
+            Objects.requireNonNull(f, "f is null");
+            return empty().addAll(Collections.tabulate(n, f));
+        }
+
+        public BitSet<T> fill(int n, Supplier<? extends T> s) {
+            Objects.requireNonNull(s, "s is null");
+            return empty().addAll(Collections.fill(n, s));
+        }
+    }
+
+    static <T> Builder<T> withRelations(Function1<Integer, T> fromInt, Function1<T, Integer> toInt) {
+        return new Builder<>(fromInt, toInt);
+    }
+
+    static <T extends Enum<T>> Builder<T> withEnum(Class<T> clz) {
```

Comment:
minor suggestion: instead of using abbreviations and misspellings of words (cla*zz*), could we please rename to reflect the intention, i.e. `Class<T> enumClass`?